go/vt/vtgate/planbuilder/gen4_planner.go

```diff
@@ -394,11 +388,82 @@ func gen4DeleteStmtPlanner(
 		return nil, err
 	}
 
-	plan, err = pushCommentDirectivesOnPlan(plan, deleteStmt)
+	plan = pushCommentDirectivesOnPlan(plan, deleteStmt)
+
+	setLockOnAllSelect(plan)
+
+	if err := plan.WireupGen4(ctx); err != nil {
+		return nil, err
+	}
+
+	return newPlanResult(plan.Primitive(), operators.TablesUsed(op)...), nil
+}
+
+func deleteUnshardedShortcut(stmt *sqlparser.Delete, ks *vindexes.Keyspace, tables []*vindexes.Table) logicalPlan {
+	edml := engine.NewDML()
+	edml.Keyspace = ks
+	edml.Table = tables
+	edml.Opcode = engine.Unsharded
+	edml.Query = generateQuery(stmt)
+	return &primitiveWrapper{prim: &engine.Delete{DML: edml}}
+}
+
+func gen4InsertStmtPlanner(version querypb.ExecuteOptions_PlannerVersion, insStmt *sqlparser.Insert, reservedVars *sqlparser.ReservedVars, vschema plancontext.VSchema) (*planResult, error) {
+	ksName := ""
+	if ks, _ := vschema.DefaultKeyspace(); ks != nil {
+		ksName = ks.Name
+	}
+	semTable, err := semantics.Analyze(insStmt, ksName, vschema)
+	if err != nil {
+		return nil, err
+	}
+	// record any warning as planner warning.
+	vschema.PlannerWarning(semTable.Warning)
+
+	err = rewriteRoutedTables(insStmt, vschema)
+	if err != nil {
+		return nil, err
+	}
+	// remove any alias added from routing table.
+	// insert query does not support table alias.
+	insStmt.Table.As = sqlparser.NewIdentifierCS("")
+
+	// Check single unsharded. Even if the table is for single unsharded but sequence table is used.
+	// We cannot shortcut here as sequence column needs additional planning.
+	ks, tables := semTable.SingleUnshardedKeyspace()
+	if ks != nil && tables[0].AutoIncrement == nil {
+		plan := insertUnshardedShortcut(insStmt, ks, tables)
+		plan = pushCommentDirectivesOnPlan(plan, insStmt)
+		return newPlanResult(plan.Primitive(), operators.QualifiedTables(ks, tables)...), nil
+	}
+
+	tblInfo, err := semTable.TableInfoFor(semTable.TableSetFor(insStmt.Table))
+	if err != nil {
+		return nil, err
+	}
+	if tblInfo.GetVindexTable().Keyspace.Sharded && semTable.NotUnshardedErr != nil {
```

Comment:
we don't need to check the tblInfo - I think the idea is that if we can't use the unshardedShortCut, we have to check this error and fail.

Reply:
we will come here for unsharded tables as well when they have vitess sequence in use. I'm not sure why it was supported at the first place. just keeping backward compatibility.
To do that and not bail on sharded error for unsharded table doing this check.